middleware/http/http_handler.go

```diff
@@ -0,0 +1,100 @@
+package http
+
+import (
+	"net/http"
+	"strconv"
+	"sync/atomic"
+
+	zipkin "github.com/openzipkin/zipkin-go"
+	"github.com/openzipkin/zipkin-go/model"
+	"github.com/openzipkin/zipkin-go/propagation/b3"
+)
+
+type httpHandler struct {
+	tracer *zipkin.Tracer
+	name   string
+	next   http.Handler
+}
+
+// WrapHTTPHandler wraps a standard http.Handler with Zipkin tracing.
+func WrapHTTPHandler(t *zipkin.Tracer, op string, h http.Handler) http.Handler {
+	return &httpHandler{
+		tracer: t,
+		next:   h,
+		name:   op,
+	}
+}
+
+// ServeHTTP implements http.Handler.
+func (h httpHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// try to extract B3 Headers from upstream
+	sc := h.tracer.Extract(b3.ExtractHTTP(r))
+
+	remoteEndpoint, _ := zipkin.NewEndpoint("", r.RemoteAddr)
+
+	// create Span using SpanContext if found
+	sp := h.tracer.StartSpan(
+		h.name,
+		zipkin.Kind(model.Server),
+		zipkin.Parent(sc),
+		zipkin.RemoteEndpoint(remoteEndpoint),
+	)
+	defer sp.Finish()
+
+	// add our span to context
+	ctx := zipkin.NewContext(r.Context(), sp)
+
+	// tag typical HTTP request items
+	zipkin.TagHTTPMethod.Set(sp, r.Method)
+	zipkin.TagHTTPUrl.Set(sp, r.URL.String())
+	zipkin.TagHTTPRequestSize.Set(sp, strconv.FormatInt(r.ContentLength, 10))
+
+	// create http.ResponseWriter interceptor for tracking response size and
+	// status code.
+	ri := &rwInterceptor{w: w, statusCode: 200}
+
+	// tag found response size and status code on exit
+	defer func() {
+		code := ri.getStatusCode()
+		sCode := strconv.Itoa(code)
+		if code > 399 {
+			zipkin.TagError.Set(sp, sCode)
+		}
+		zipkin.TagHTTPStatusCode.Set(sp, sCode)
+		zipkin.TagHTTPResponseSize.Set(sp, ri.getResponseSize())
+	}()
+
+	// call next http Handler func using our updated context.
+	h.next.ServeHTTP(ri, r.WithContext(ctx))
+}
+
+// rwInterceptor intercepts the ResponseWriter so it can track response size
```

Comment:
most instrumentation don't add response size at the moment.. you'll notice if blocking on that it can delay the span (you can also report tags after the fact though). Most typically, the server span ends when headers are sent. Not saying this is correct, just a tendency.. either way, summarizing what behavior is normal is good.

In brave, for example, we add very minimum tags (though people can choose to add more, which here would be a separate feature for later I suspect) https://github.com/openzipkin/brave/tree/master/instrumentation/http#span-data-policy

Reply:
I think I’ll make this optional logic. Given this is middleware (self contained layer of an onion) i prefer to keep logic as is if activated by consumer